File: web-react/src/tamizaje/motorTamizaje.test.js
/**
 * Suite de conformidad del motor de tamizaje (JavaScript).
 *
 * No usa framework de pruebas a proposito: corre con `node motorTamizaje.test.js`
 * sin instalar nada. Si el equipo ya tiene vitest o jest, los mismos vectores se
 * pueden envolver en `it()` sin tocar el motor.
 *
 * Lee los casos de ../compartido/vectores_conformidad.json, el MISMO archivo que
 * usan las suites de Python y Kotlin.
 */

import { existsSync, readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join, parse } from "node:path";

import {
  evaluarCaso,
  evaluarRonda,
  bandaPorAltitud,
  validarEntrada,
  BANDAS,
  VERSION_UMBRALES,
  Resultado,
} from "./motorTamizaje.js";

/**
 * Busca compartido/vectores_conformidad.json hacia la raiz del repo, para que la
 * prueba funcione igual si el archivo se mueve de carpeta.
 */
function rutaDeVectores() {
  let dir = dirname(fileURLToPath(import.meta.url));
  const raiz = parse(dir).root;
  while (true) {
    const candidata = join(dir, "compartido", "vectores_conformidad.json");
    if (existsSync(candidata)) return candidata;
    if (dir === raiz) break;
    dir = dirname(dir);
  }
  throw new Error(
    "No se encontro compartido/vectores_conformidad.json subiendo desde " +
      dirname(fileURLToPath(import.meta.url))
  );
}

const spec = JSON.parse(readFileSync(rutaDeVectores(), "utf-8"));

let pasaron = 0;
const fallos = [];

function comprobar(nombre, condicion, detalle) {
  if (condicion) pasaron++;
  else fallos.push(`${nombre}: ${detalle}`);
}

// --- 1. Vectores de conformidad compartidos -------------------------------

comprobar(
  "version de umbrales",
  spec.version_umbrales === VERSION_UMBRALES,
  `los vectores son de ${spec.version_umbrales} y el motor de ${VERSION_UMBRALES}`
);

for (const caso of spec.casos) {
  const r = evaluarCaso(caso.entrada);
  const e = caso.espera;

  comprobar(caso.id, r.ok === true, `la entrada no valido: ${JSON.stringify(r.errores)}`);
  if (!r.ok) continue;

  comprobar(caso.id, r.resultado === e.resultado, `resultado ${r.resultado}, se esperaba ${e.resultado}`);

  if (e.banda !== undefined) {
    comprobar(caso.id, r.banda?.id === e.banda, `banda ${r.banda?.id}, se esperaba ${e.banda}`);
  }
  if (e.motivoNoElegible !== undefined) {
    comprobar(caso.id, r.motivoNoElegible === e.motivoNoElegible, `motivo ${r.motivoNoElegible}, se esperaba ${e.motivoNoElegible}`);
  }
  if (e.proximaRonda !== undefined) {
    comprobar(caso.id, r.proximaRonda === e.proximaRonda, `proximaRonda ${r.proximaRonda}, se esperaba ${e.proximaRonda}`);
  }
  if (e.diferenciaSpo2 !== undefined) {
    comprobar(caso.id, r.diferenciaSpo2 === e.diferenciaSpo2, `diferencia ${r.diferenciaSpo2}, se esperaba ${e.diferenciaSpo2}`);
  }
  if (e.avisoPresente !== undefined) {
    const hay = r.avisos.some((a) => a.codigo === e.avisoPresente);
    comprobar(caso.id, hay, `falta el aviso ${e.avisoPresente}`);
  }
  if (e.avisoAusente !== undefined) {
    const hay = r.avisos.some((a) => a.codigo === e.avisoAusente);
    comprobar(caso.id, !hay, `sobra el aviso ${e.avisoAusente}`);
  }

  // Invariante que aplica a TODOS los casos: la advertencia nunca falta.
  comprobar(caso.id, typeof r.advertencia === "string" && r.advertencia.length > 0, "falta la advertencia obligatoria");
}

// --- 2. Regresion del bug mas importante ----------------------------------
// Aplicar el umbral de nivel del mar en la sierra. Un RN sano en Juliaca
// satura ~88%: con el corte de banda 1 sale positivo, con el correcto no.

comprobar(
  "regresion_altitud",
  evaluarRonda(BANDAS[2], 88, 87) === Resultado.NEGATIVO,
  "banda 3 deberia dar negativo con 88/87"
);
comprobar(
  "regresion_altitud",
  evaluarRonda(BANDAS[0], 88, 87) === Resultado.POSITIVO,
  "banda 1 deberia dar positivo con la MISMA medicion"
);

// --- 3. Umbrales provisionales marcados -----------------------------------

comprobar("estado_umbrales", bandaPorAltitud(150).estado === "verificado", "B1 deberia estar verificada");
comprobar("estado_umbrales", bandaPorAltitud(3000).estado === "provisional", "B2 deberia estar marcada provisional");
comprobar("estado_umbrales", bandaPorAltitud(4000).estado === "provisional", "B3 deberia estar marcada provisional");

// --- 4. Validacion de entrada ---------------------------------------------

comprobar("validacion", Object.keys(validarEntrada({ altitudMsnm: 150, spo2Preductal: 98 })).length === 0, "una entrada valida no deberia dar errores");
comprobar("validacion", "spo2Preductal" in validarEntrada({ altitudMsnm: 150, spo2Preductal: 120 }), "SpO2 de 120 deberia rechazarse");
comprobar("validacion", "spo2Preductal" in validarEntrada({ altitudMsnm: 150 }), "falta SpO2 preductal, deberia ser obligatoria");
comprobar("validacion", "altitudMsnm" in validarEntrada({ altitudMsnm: 6000, spo2Preductal: 98 }), "6000 msnm esta fuera de las bandas");
comprobar("validacion", evaluarCaso({ altitudMsnm: 150, spo2Preductal: 120 }).ok === false, "evaluarCaso deberia devolver ok:false con entrada invalida");

// --- Resumen ---------------------------------------------------------------

console.log(`\n  ${pasaron} comprobaciones pasaron, ${fallos.length} fallaron`);
if (fallos.length) {
  console.log("\n  FALLOS:");
  for (const f of fallos) console.log("   -", f);
  process.exit(1);
}
console.log("  Motor JS conforme.\n");

File: app-kotlin/tamizaje/MotorTamizajeTest.kt
package pe.cardioalerta.tamizaje

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Suite de conformidad del motor de tamizaje (Kotlin).
 *
 * Va en `app/src/test/java/pe/cardioalerta/tamizaje/` — es una prueba unitaria
 * normal de JVM, no necesita emulador ni dispositivo. Corre con:
 *
 *     ./gradlew test
 *
 * Los casos vienen de VectoresConformidad.kt, generado desde el MISMO archivo
 * JSON que usan las suites de JavaScript y Python. Si un caso falla aca pero
 * pasa alla, la app y el backend estan clasificando distinto al mismo bebe.
 */
class MotorTamizajeTest {

    private fun salidaDe(entrada: MotorTamizaje.Entrada): MotorTamizaje.Salida {
        val evaluacion = MotorTamizaje.evaluarCaso(entrada)
        assertTrue(
            "La entrada no valido: $evaluacion",
            evaluacion is MotorTamizaje.Evaluacion.Valida,
        )
        return (evaluacion as MotorTamizaje.Evaluacion.Valida).salida
    }

    // --- 1. Vectores de conformidad compartidos ---------------------------

    @Test
    fun versionDeUmbralesCoincideConLosVectores() {
        assertEquals(VectoresConformidad.VERSION_UMBRALES, MotorTamizaje.VERSION_UMBRALES)
    }

    @Test
    fun todosLosVectoresDeConformidadPasan() {
        val fallos = mutableListOf<String>()

        for (caso in VectoresConformidad.CASOS) {
            val evaluacion = MotorTamizaje.evaluarCaso(caso.entrada)
            if (evaluacion !is MotorTamizaje.Evaluacion.Valida) {
                fallos += "${caso.id}: la entrada no valido"
                continue
            }
            val s = evaluacion.salida

            val resultado = s.resultado.name.lowercase()
            if (resultado != caso.resultado) {
                fallos += "${caso.id}: resultado $resultado, se esperaba ${caso.resultado}"
            }
            caso.banda?.let {
                if (s.banda?.id != it) fallos += "${caso.id}: banda ${s.banda?.id}, se esperaba $it"
            }
            caso.motivoNoElegible?.let {
                val motivo = s.motivoNoElegible?.name?.lowercase()
                if (motivo != it) fallos += "${caso.id}: motivo $motivo, se esperaba $it"
            }
            caso.proximaRonda?.let {
                if (s.proximaRonda != it) fallos += "${caso.id}: proximaRonda ${s.proximaRonda}, se esperaba $it"
            }
            caso.diferenciaSpo2?.let {
                if (s.diferenciaSpo2 != it) fallos += "${caso.id}: diferencia ${s.diferenciaSpo2}, se esperaba $it"
            }
            caso.avisoPresente?.let { codigo ->
                if (s.avisos.none { it.codigo == codigo }) fallos += "${caso.id}: falta el aviso $codigo"
            }
            caso.avisoAusente?.let { codigo ->
                if (s.avisos.any { it.codigo == codigo }) fallos += "${caso.id}: sobra el aviso $codigo"
            }
            if (s.advertencia.isBlank()) {
                fallos += "${caso.id}: falta la advertencia obligatoria"
            }
        }

        assertTrue(
            "Fallaron ${fallos.size} de ${VectoresConformidad.CASOS.size} vectores:\n" +
                fallos.joinToString("\n"),
            fallos.isEmpty(),
        )
    }

    // --- 2. Regresion del bug mas importante ------------------------------

    /**
     * Aplicar el umbral de nivel del mar en la sierra. Un recien nacido sano en
     * Juliaca (3825 msnm) satura alrededor de 88%: con el corte de banda 1 sale
     * positivo, con el de su banda no. Este es el caso que justifica el proyecto.
     */
    @Test
    fun altitudAltaNoMarcaPositivoAUnNeonatoSanoDeAltura() {
        val b1 = MotorTamizaje.bandaPorAltitud(150)!!
        val b3 = MotorTamizaje.bandaPorAltitud(3825)!!
        assertEquals(MotorTamizaje.Resultado.NEGATIVO, MotorTamizaje.evaluarRonda(b3, 88, 87))
        assertEquals(MotorTamizaje.Resultado.POSITIVO, MotorTamizaje.evaluarRonda(b1, 88, 87))
    }

    // --- 3. La puerta de elegibilidad -------------------------------------

    @Test
    fun recienNacidoSintomaticoNoSeTamiza() {
        val s = salidaDe(
            MotorTamizaje.Entrada(
                altitudMsnm = 150, spo2Preductal = 88, spo2Postductal = 86,
                horasDeVida = 24.0, sintomas = setOf("cianosis_central"),
            )
        )
        assertEquals(MotorTamizaje.Resultado.NO_ELEGIBLE, s.resultado)
        assertEquals(MotorTamizaje.MotivoNoElegible.SINTOMATICO, s.motivoNoElegible)
        assertTrue(s.sintomasDeAlarma.contains("Cianosis central"))
    }

    @Test
    fun elSintomaticoDominaSobreLasDemasCausas() {
        // Bebe de 10 h con mala perfusion: el motivo es sintomatico, no la edad.
        // Importa porque cambia la conducta de "espere" a "evalue ahora".
        assertEquals(
            MotorTamizaje.MotivoNoElegible.SINTOMATICO,
            MotorTamizaje.evaluarElegibilidad(10.0, setOf("mala_perfusion"), true, true),
        )
    }

    @Test
    fun elSoploNoExcluyeDelTamizajePeroSiAvisa() {
        val s = salidaDe(
            MotorTamizaje.Entrada(
                altitudMsnm = 150, spo2Preductal = 98, spo2Postductal = 97,
                horasDeVida = 30.0, sintomas = setOf("soplo_cardiaco"),
            )
        )
        assertEquals(MotorTamizaje.Resultado.NEGATIVO, s.resultado)
        assertTrue(s.avisos.any { it.codigo == "soplo_cardiaco" })
    }

    @Test
    fun elegibilidadNormal() {
        assertNull(MotorTamizaje.evaluarElegibilidad(30.0, emptySet(), false, false))
        assertEquals(
            MotorTamizaje.MotivoNoElegible.MENOR_24H,
            MotorTamizaje.evaluarElegibilidad(23.9, emptySet(), false, false),
        )
        assertNull(MotorTamizaje.evaluarElegibilidad(24.0, emptySet(), false, false))
    }

    // --- 4. Medicion incompleta -------------------------------------------

    @Test
    fun sinPostductalElTamizajeQuedaIncompletoNoNegativo() {
        val s = salidaDe(
            MotorTamizaje.Entrada(altitudMsnm = 150, spo2Preductal = 98, horasDeVida = 30.0)
        )
        assertEquals(MotorTamizaje.Resultado.INCOMPLETO, s.resultado)
        assertTrue(s.avisos.any { it.codigo == "falta_postductal" })
    }

    @Test
    fun sinPostductalPeroConPreductalCriticoSiEsPositivo() {
        // No hace falta el pie para reprobar a un bebe que satura 85 a nivel del mar.
        val s = salidaDe(
            MotorTamizaje.Entrada(altitudMsnm = 150, spo2Preductal = 85, horasDeVida = 30.0)
        )
        assertEquals(MotorTamizaje.Resultado.POSITIVO, s.resultado)
    }

    // --- 5. Retamizaje -----------------------------------------------------

    @Test
    fun elRetamizajeCierraEnLaTerceraRonda() {
        val entrada = MotorTamizaje.Entrada(
            altitudMsnm = 150, spo2Preductal = 92, spo2Postductal = 91, horasDeVida = 30.0,
        )
        assertEquals(MotorTamizaje.Resultado.REPETIR, salidaDe(entrada.copy(ronda = 1)).resultado)
        assertEquals(MotorTamizaje.Resultado.REPETIR, salidaDe(entrada.copy(ronda = 2)).resultado)
        assertEquals(MotorTamizaje.Resultado.POSITIVO, salidaDe(entrada.copy(ronda = 3)).resultado)
    }

    @Test
    fun elRepetirTraeLaProximaRondaYLaEspera() {
        val s = salidaDe(
            MotorTamizaje.Entrada(
                altitudMsnm = 150, spo2Preductal = 92, spo2Postductal = 91,
                horasDeVida = 30.0, ronda = 1,
            )
        )
        assertEquals(2, s.proximaRonda)
        assertEquals(60, s.minutosEspera)
    }

    // --- 6. Diferencial ----------------------------------------------------

    @Test
    fun laDiferenciaConservaElSignoParaElRegistro() {
        assertEquals(-5, MotorTamizaje.diferenciaConSigno(92, 97))
        assertEquals(6, MotorTamizaje.diferenciaConSigno(96, 90))
        assertNull(MotorTamizaje.diferenciaConSigno(96, null))
    }

    // --- 7. Umbrales provisionales marcados --------------------------------

    @Test
    fun losUmbralesProvisionalesEstanMarcadosYAvisan() {
        assertEquals(MotorTamizaje.EstadoUmbral.VERIFICADO, MotorTamizaje.bandaPorAltitud(150)!!.estado)
        assertEquals(MotorTamizaje.EstadoUmbral.PROVISIONAL, MotorTamizaje.bandaPorAltitud(3000)!!.estado)
        assertEquals(MotorTamizaje.EstadoUmbral.PROVISIONAL, MotorTamizaje.bandaPorAltitud(4000)!!.estado)

        val s = salidaDe(
            MotorTamizaje.Entrada(
                altitudMsnm = 3825, spo2Preductal = 90, spo2Postductal = 89, horasDeVida = 30.0,
            )
        )
        assertTrue(s.avisos.any { it.codigo == "umbral_provisional" })
    }

    // --- 8. Los signos vitales no alteran el resultado ----------------------

    @Test
    fun losSignosVitalesAvisanPeroNoCambianElResultado() {
        val base = MotorTamizaje.Entrada(
            altitudMsnm = 150, spo2Preductal = 98, spo2Postductal = 97, horasDeVida = 30.0,
        )
        val sinSignos = salidaDe(base)
        val conSignos = salidaDe(base.copy(fcLpm = 200, frRpm = 80, pesoKg = 2.0))

        assertEquals(sinSignos.resultado, conSignos.resultado)
        assertTrue(conSignos.avisos.any { it.codigo == "fc_alta" })
        assertTrue(conSignos.avisos.any { it.codigo == "fr_alta" })
        assertTrue(conSignos.avisos.any { it.codigo == "bajo_peso" })
    }

    // --- 9. Validacion de entrada ------------------------------------------

    @Test
    fun laEntradaInvalidaSeRechaza() {
        val evaluacion = MotorTamizaje.evaluarCaso(
            MotorTamizaje.Entrada(altitudMsnm = 150, spo2Preductal = 120)
        )
        assertTrue(evaluacion is MotorTamizaje.Evaluacion.Invalida)
        assertNotNull((evaluacion as MotorTamizaje.Evaluacion.Invalida).errores["spo2Preductal"])
    }

    @Test
    fun laAltitudFueraDeBandasSeRechaza() {
        val evaluacion = MotorTamizaje.evaluarCaso(
            MotorTamizaje.Entrada(altitudMsnm = 6000, spo2Preductal = 98)
        )
        assertTrue(evaluacion is MotorTamizaje.Evaluacion.Invalida)
    }

    // --- 10. La advertencia nunca falta ------------------------------------

    @Test
    fun laAdvertenciaSiempreAcompanaAlResultado() {
        for (caso in VectoresConformidad.CASOS) {
            val evaluacion = MotorTamizaje.evaluarCaso(caso.entrada)
            if (evaluacion is MotorTamizaje.Evaluacion.Valida) {
                assertTrue(
                    "Caso ${caso.id} salio sin advertencia",
                    evaluacion.salida.advertencia.isNotBlank(),
                )
            }
        }
    }
}
